### PR TITLE
[search] CategoriesInfo.

### PR DESCRIPTION
**Pull request by ygorshenin**

### CODE REVIEW

File: search/ranking_utils.hpp [hunk 1]
```diff
@@ -17,9 +17,36 @@
...
+class TokenSlice;
+
+class CategoriesInfo
+{
+public:
+  CategoriesInfo(feature::TypesHolder const & holder, TokenSlice const & tokens,
+                 Locales const & locales, CategoriesHolder const & categories);
+
+  // Returns true when all tokens correspond to categories in
+  // |holder|.
```
Comment (mpimenov) [inline search/ranking_utils.hpp]: It's not clear what |holder| is here. Probably add a local copy, |m_holder|?

File: search/ranking_utils.hpp [hunk 2]
```diff
@@ -17,9 +17,36 @@
...
+  bool IsPureCategories() const { return m_pureCategories; }
+
+  // Returns true when all tokens are categories tokens but do not
```
Comment (mpimenov) [inline search/ranking_utils.hpp]: s/do not correspond/none of them corresponds/
for clarity

File: search/ranker.hpp [hunk 3]
```diff
@@ -20,10 +20,11 @@
...
 
-#include "std/set.hpp"
-#include "std/string.hpp"
-#include "std/utility.hpp"
-#include "std/vector.hpp"
+#include <cstdint>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
```
Comment (mpimenov) [inline search/ranker.hpp]: If you do this, add the std:: prefixes where they belong, please.

File: search/ranking_utils.cpp [hunk 4]
```diff
@@ -1,21 +1,60 @@
...
+struct TokenInfo
+{
+  bool m_isCategoryToken = false;
+  bool m_inFeatureTypes = false;
+};
+
 UniString AsciiToUniString(char const * s) { return UniString(s, s + strlen(s)); }
 }  // namespace
 
+// CategoriesInfo ----------------------------------------------------------------------------------
```
Comment (mpimenov) [inline search/ranking_utils.cpp]: What about a separating comment for standalone functions?

File: search/ranker.hpp (hunk 3 above)
Comment (ygorshenin) [reply search/ranker.hpp]: Done

File: search/ranking_utils.hpp (hunk 2 above)
Comment (ygorshenin) [reply search/ranking_utils.hpp]: Done

File: search/ranking_utils.hpp (hunk 1 above)
Comment (ygorshenin) [reply search/ranking_utils.hpp]: I'd prefer not to hold reference to TypesHolder.

File: search/ranking_utils.hpp [hunk 5]
```diff
@@ -17,9 +17,36 @@
...
 namespace search
 {
 class QueryParams;
+class TokenSlice;
+
+class CategoriesInfo
+{
+public:
+  CategoriesInfo(feature::TypesHolder const & holder, TokenSlice const & tokens,
+                 Locales const & locales, CategoriesHolder const & categories);
```
[the remark this answers is not in the document]
Comment (ygorshenin) [reply search/ranking_utils.hpp]: Done

File: search/ranking_info.hpp [hunk 6]
```diff
@@ -1,11 +1,10 @@
 #pragma once
 
+#include "search/common.hpp"
 #include "search/model.hpp"
 #include "search/pre_ranking_info.hpp"
 #include "search/ranking_utils.hpp"
 
-#include "std/iostream.hpp"
```
Comment (tatiana-kondakova) [inline search/ranking_info.hpp]: `#include <iostream>` ?

File: search/ranking_utils.hpp [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 #include "search/model.hpp"
 #include "search/query_params.hpp"
-#include "search/utils.hpp"
+#include "search/common.hpp"
```
Comment (mpimenov) [inline search/ranking_utils.hpp]: sort, please

File: search/ranking_info.hpp (hunk 6 above)
Comment (ygorshenin) [reply search/ranking_info.hpp]: As you can see, it was removed.

File: search/ranking_utils.cpp (hunk 4 above)
Comment (ygorshenin) [reply search/ranking_utils.cpp]: Done

File: search/ranking_utils.hpp (hunk 7 above)
Comment (ygorshenin) [reply search/ranking_utils.hpp]: Done

File: search/ranking_utils.cpp [hunk 8]
```diff
@@ -1,21 +1,61 @@
...
+
+    info.m_isCategoryToken = true;
+    if (holder.Has(t))
+      info.m_inFeatureTypes = true;
+  });
+
+  // Note that m_inFeatureTypes implies m_isCategoryToken.
+
+  m_pureCategories = all_of(infos.begin(), infos.end(), [](TokenInfo const & info) {
+    ASSERT(!info.m_inFeatureTypes || info.m_isCategoryToken, ());
```
Comment (tatiana-kondakova) [inline search/ranking_utils.cpp]: up to you, но кажется
`!(info.m_inFeatureTypes && !info.m_isCategoryToken)`
понятнее

File: search/ranker.hpp [hunk 9]
```diff
@@ -91,26 +94,28 @@ class Ranker
 
   void SuggestStrings();
 
-  virtual void SetPreRankerResults(vector<PreRankerResult> && preRankerResults)
+  virtual void SetPreRankerResults(std::vector<PreRankerResult> && preRankerResults)
   {
     m_preRankerResults = move(preRankerResults);
```
Comment (tatiana-kondakova) [inline search/ranker.hpp]: `std::move`

File: search/ranking_utils.cpp (hunk 8 above)
Comment (ygorshenin) [reply search/ranking_utils.cpp]: ```a => b``` is equivalent to ```~a or b```. Of course, it's equivalent to ```~(a and ~b)```, but the former transformation is simpler and well known.

File: search/ranker.hpp (hunk 9 above)
Comment (ygorshenin) [reply search/ranker.hpp]: Done

File: search/ranking_info.hpp (hunk 6 above)
Comment (tatiana-kondakova) [reply search/ranking_info.hpp]: lines 40-42:
```C++
static void PrintCSVHeader(ostream & os);
void ToCSV(ostream & os) const;
```

File: search/ranking_utils.cpp (hunk 8 above)
Comment (tatiana-kondakova) [reply search/ranking_utils.cpp]: ok

File: search/ranking_info.hpp (hunk 6 above)
Comment (ygorshenin) [reply search/ranking_info.hpp]: Looks like this file was touched occasionally :) Fixed anyway.

**Pull request merged**